keychain_win.cpp

```diff
@@ -0,0 +1,211 @@
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+// clang-format off
+// make sure windows.h is included before wincred.h
+#include "keychain.h"
+
+#include <memory>
+
+#define UNICODE
+
+#include <windows.h>
+#include <wincred.h>
+#include <intsafe.h> // for DWORD_MAX
+// clang-format on
+
+namespace {
+
+static DWORD cred_type = CRED_TYPE_GENERIC;
+
+struct LpwstrDeleter {
+    void operator()(WCHAR *p) const { delete[] p; }
+};
+using ScopedLpwstr = std::unique_ptr<WCHAR, LpwstrDeleter>;
+
+LPWSTR utf8ToWideChar(const std::string &utf8) {
+    int len = MultiByteToWideChar(
+        CP_UTF8,
+        0, // flags must be 0 for UTF-8
+        utf8.c_str(),
+        -1,      // rely on null-terminated input string
+        nullptr, // no out-buffer needed
+        0); // return required buffer size rather than writing to an out-buffer
+    if (len == 0) {
+        return nullptr;
+    }
+
+    LPWSTR buffer = new WCHAR[len];
+    int bytesWritten =
+        MultiByteToWideChar(CP_UTF8, 0, utf8.c_str(), -1, buffer, len);
+
+    if (bytesWritten == 0) {
+        // failure; cleanup buffer manually
+        delete[] buffer;
+        return nullptr;
+    }
+
+    return buffer;
+}
+
+std::string wideCharToAnsi(LPWSTR wChar) {
+    std::string result;
+    if (wChar == nullptr) {
+        return result;
+    }
+
+    int len =
+        WideCharToMultiByte(CP_ACP, 0, wChar, -1, nullptr, 0, nullptr, nullptr);
+    if (len == 0) {
+        return result;
+    }
+
+    std::unique_ptr<char[]> buffer(new char[len]);
+    int bytesWritten = WideCharToMultiByte(
+        CP_ACP, 0, wChar, -1, buffer.get(), len, nullptr, nullptr);
+
+    if (bytesWritten != 0) {
+        result = std::string(buffer.get());
+    }
+
+    return result;
+}
+
+std::string getErrorMessage(DWORD errorCode) {
+    std::string errMsg;
+    LPWSTR errBuffer = nullptr;
+    ::FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
+                    nullptr, // ignored for the flags we use
+                    errorCode,
+                    0, // figure out LANGID automatically
+                    reinterpret_cast<LPWSTR>(&errBuffer),
```

Comment:
Since we `#define UNICODE` on the top of this file this is just a `typedef LPWSTR LPTSTR`. I don't know what's considered good style here, but I'm sure we rely on this in most of the string conversion functions above. So it might even be safer to go with this, in case someone drops the `#define UNICODE`.